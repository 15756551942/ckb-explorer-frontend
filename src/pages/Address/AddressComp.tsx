import React, { ReactNode, useContext, useState } from 'react'
import Pagination from '../../components/Pagination'
import HelpIcon from '../../assets/qa_help.png'
import OverviewCard, { OverviewItemData } from '../../components/Card/OverviewCard'
import TitleCard from '../../components/Card/TitleCard'
import Tooltip from '../../components/Tooltip'
import TransactionItem from '../../components/TransactionItem/index'
import { AppContext } from '../../contexts/providers/index'
import i18n from '../../utils/i18n'
import { localeNumberString } from '../../utils/number'
import { isMobile } from '../../utils/screen'
import { adaptMobileEllipsis, adaptPCEllipsis } from '../../utils/string'
import { shannonToCkb } from '../../utils/util'
import {
  AddressLockScriptItemPanel,
  AddressLockScriptPanel,
  AddressPendingRewardTitlePanel,
  AddressTransactionsPagination,
  AddressTransactionsPanel,
} from './styled'
import browserHistory from '../../routes/history'
import DecimalCapacity from '../../components/DecimalCapacity'
import { parseSimpleDateNoSecond } from '../../utils/date'

const addressContent = (address: string) => {
  const addressText = isMobile() ? adaptMobileEllipsis(address, 10) : adaptPCEllipsis(address, 13, 50)
  return address ? addressText : i18n.t('address.unable_decode_address')
}

const AddressPendingRewardTitle = () => {
  const [show, setShow] = useState(false)
  return (
    <AddressPendingRewardTitlePanel>
      {`${i18n.t('address.pending_reward')}`}
      <div
        id="address__pending_reward_help"
        tabIndex={-1}
        onFocus={() => {}}
        onMouseOver={() => {
          setShow(true)
          const p = document.querySelector('.page') as HTMLElement
          if (p) {
            p.setAttribute('tabindex', '-1')
          }
        }}
        onMouseLeave={() => {
          setShow(false)
          const p = document.querySelector('.page') as HTMLElement
          if (p) {
            p.removeAttribute('tabindex')
          }
        }}
      >
        <img src={HelpIcon} alt="Pending Reward Help" />
      </div>
      <Tooltip show={show} targetElementId="address__pending_reward_help">
        {i18n.t('address.pending_reward_tooltip')}
      </Tooltip>
    </AddressPendingRewardTitlePanel>
  )
}

const AddressLockScriptItem = ({ title, children }: { title: string; children?: ReactNode }) => {
  return (
    <AddressLockScriptItemPanel>
      <div className="address_lock_script__title">
        <span>{title}</span>
      </div>
      <div className="address_lock_script__content">{children}</div>
    </AddressLockScriptItemPanel>
  )
}

const AddressLockScript = ({ script }: { script: State.Script }) => {
  return (
    <AddressLockScriptPanel>
      <div className="address__lock_script_title">{i18n.t('address.lock_script')}</div>
      <AddressLockScriptItem title={i18n.t('address.code_hash')}>
        <span>{script.codeHash}</span>
      </AddressLockScriptItem>
      <AddressLockScriptItem title={i18n.t('address.args')}>
        <span>{script.args}</span>
      </AddressLockScriptItem>
      <AddressLockScriptItem title={i18n.t('address.hash_type')}>
        <code>{script.hashType}</code>
      </AddressLockScriptItem>
    </AddressLockScriptPanel>
  )
}

const getAddressInfo = (addressState: State.AddressState) => {
  const { address } = addressState
  const items: OverviewItemData[] = [
    {
      title: i18n.t('address.balance'),
      content: <DecimalCapacity value={localeNumberString(shannonToCkb(address.balance))} />,
    },
    {
      title: i18n.t('transaction.transactions'),
      content: localeNumberString(address.transactionsCount),
    },
    {
      title: i18n.t('address.dao_deposit'),
      content: <DecimalCapacity value={localeNumberString(shannonToCkb(address.daoDeposit))} />,
    },
    {
      title: i18n.t('address.compensation'),
      content: <DecimalCapacity value={localeNumberString(shannonToCkb(address.interest))} />,
    },
  ]

  if (address.pendingRewardBlocksCount) {
    items.push({
      title: <AddressPendingRewardTitle />,
      content: `${address.pendingRewardBlocksCount} ${address.pendingRewardBlocksCount > 1 ? 'blocks' : 'block'}`,
    })
  }
  if (address.type === 'LockHash' && address) {
    items.push({
      title: i18n.t('address.address'),
      content: addressContent(address.addressHash),
    })
  }
  const { lockInfo } = address
  if (lockInfo && lockInfo.epochNumber) {
    items.push({
      title: i18n.t('address.lock_until'),
      content: `${lockInfo.epochNumber} ${i18n.t('address.epoch')} (~ ${parseSimpleDateNoSecond(
        lockInfo.estimatedUnlockTime,
      )})`,
    })
  }

  return items
}

export const AddressOverview = () => {
  const { addressState } = useContext(AppContext)
  return (
    <>
      <TitleCard title={i18n.t('common.overview')} />
      <OverviewCard items={getAddressInfo(addressState)}>
        {addressState && addressState.address && addressState.address.lockScript && (
          <AddressLockScript script={addressState.address.lockScript} />
        )}
      </OverviewCard>
    </>
  )
}

export const AddressTransactions = ({
  currentPage,
  pageSize,
  address,
}: {
  currentPage: number
  pageSize: number
  address: string
}) => {
  const { addressState, app } = useContext(AppContext)
  const { tipBlockNumber } = app
  const { transactions = [] } = addressState

  const totalPages = Math.ceil(addressState.total / pageSize)

  const onChange = (page: number) => {
    browserHistory.replace(`/address/${address}?page=${page}&size=${pageSize}`)
  }

  return (
    <>
      {transactions.length > 0 && <TitleCard title={i18n.t('transaction.transactions')} />}
      <AddressTransactionsPanel>
        {addressState.transactions.map((transaction: State.Transaction, index: number) => {
          return (
            transaction && (
              <TransactionItem
                address={addressState.address.addressHash}
                transaction={transaction}
                confirmation={tipBlockNumber - transaction.blockNumber + 1}
                key={transaction.transactionHash}
                isLastItem={index === addressState.transactions.length - 1}
              />
            )
          )
        })}
      </AddressTransactionsPanel>
      {totalPages > 1 && (
        <AddressTransactionsPagination>
          <Pagination currentPage={currentPage} totalPages={totalPages} onChange={onChange} />
        </AddressTransactionsPagination>
      )}
    </>
  )
}

export default {
  AddressOverview,
  AddressTransactions,
}
