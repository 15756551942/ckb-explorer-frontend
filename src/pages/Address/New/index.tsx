import queryString from 'query-string'
import React, { ReactNode, useContext, useEffect, useReducer } from 'react'
import { RouteComponentProps } from 'react-router-dom'
import Pagination from 'rc-pagination'
import localeInfo from 'rc-pagination/lib/locale/en_US'
import CopyIcon from '../../../assets/copy.png'
import ItemPointIcon from '../../../assets/item_point.png'
import { Tooltip } from '../../../components/Label'
import AppContext from '../../../contexts/App'
import { fetchAddressInfo, fetchTipBlockNumber, fetchTransactionsByAddress } from '../../../service/http/fetcher'
import i18n from '../../../utils/i18n'
import { localeNumberString } from '../../../utils/number'
import { isMobile } from '../../../utils/screen'
import { parsePageNumber, startEndEllipsis } from '../../../utils/string'
import { copyElementValue, shannonToCkb } from '../../../utils/util'
import TransactionCard from '../../../components/Transaction/TransactionCard/index'
import TransactionItem from '../../../components/Transaction/TransactionItem/index'
import {
  AddressHashPanel,
  AddressLockScriptItemPanel,
  AddressOverviewContentPanel,
  AddressOverviewItemPanel,
  AddressOverviewPanel,
  AddressContentPanel,
  AddressTitlePanel,
  AddressTransactionsPanel,
  AddressTransactionsPagition,
} from './styled'

enum PageParams {
  PageNo = 1,
  PageSize = 10,
  MaxPageSize = 100,
}

const initAddress: State.Address = {
  address_hash: '',
  lock_hash: '',
  balance: 0,
  transactions_count: 0,
  pending_reward_blocks_count: 0,
  cell_consumed: 0,
  lock_script: {
    args: [],
    code_hash: '',
  },
}

const Actions = {
  address: 'ADDRESS',
  transactions: 'TRANSACTIONS',
  total: 'TOTAL',
  tipBlockNumber: 'TIP_BLOCK_NUMBER',
}

const reducer = (state: any, action: any) => {
  switch (action.type) {
    case Actions.address:
      return {
        ...state,
        address: action.payload.address,
      }
    case Actions.transactions:
      return {
        ...state,
        transactions: action.payload.transactions,
      }
    case Actions.total:
      return {
        ...state,
        total: action.payload.total,
      }
    case Actions.tipBlockNumber:
      return {
        ...state,
        tipBlockNumber: action.payload.tipBlockNumber,
      }
    default:
      return state
  }
}

const getAddressInfo = (hash: string, dispatch: any) => {
  fetchAddressInfo(hash).then(response => {
    const { data } = response as Response.Response<Response.Wrapper<State.Address>>
    if (data) {
      dispatch({
        type: Actions.address,
        payload: {
          address: data.attributes,
        },
      })
    }
  })
}

const getTransactions = (hash: string, page: number, size: number, dispatch: any) => {
  fetchTransactionsByAddress(hash, page, size).then(response => {
    const { data, meta } = response as Response.Response<Response.Wrapper<State.Transaction>[]>
    if (data) {
      dispatch({
        type: Actions.transactions,
        payload: {
          transactions: data,
        },
      })
    }
    if (meta) {
      dispatch({
        type: Actions.total,
        payload: {
          total: meta.total,
        },
      })
    }
  })
}

const getTipBlockNumber = (dispatch: any) => {
  fetchTipBlockNumber().then(response => {
    const { data } = response as Response.Response<Response.Wrapper<State.Statistics>>
    if (data) {
      dispatch({
        type: Actions.tipBlockNumber,
        payload: {
          tipBlockNumber: parseInt(data.attributes.tip_block_number, 10),
        },
      })
    }
  })
}

export const PendingRewardTooltip: Tooltip = {
  tip: i18n.t('address.pending_reward_tooltip'),
  haveHelpIcon: true,
  offset: 0.7,
}

const addressContent = (address: State.Address) => {
  const addressText = isMobile() ? startEndEllipsis(address.address_hash, 10) : address.address_hash
  return address.address_hash ? addressText : i18n.t('address.unable_decode_address')
}

const initialState = {
  address: initAddress,
  transactions: [] as Response.Wrapper<State.Transaction>[],
  total: 1,
  tipBlockNumber: 0,
}

const AddressOverviewItem = ({ title, content }: { title: string; content: ReactNode }) => {
  return (
    <AddressOverviewItemPanel>
      <div className="address_overview_item__title">{title}</div>
      <div className="address_overview_item__value">{content}</div>
    </AddressOverviewItemPanel>
  )
}

interface AddressOverviewItem {
  title: string
  content: ReactNode
}

const AddressOverviewContent = ({ items }: { items: AddressOverviewItem[] }) => {
  const leftItems: AddressOverviewItem[] = []
  const rightItems: AddressOverviewItem[] = []
  items.forEach((item, index) => {
    if (index % 2 === 0) {
      leftItems.push(item)
    } else {
      rightItems.push(item)
    }
  })
  return (
    <AddressOverviewContentPanel>
      <div className="address_overview_content__left_items">
        {leftItems.map(item => (
          <AddressOverviewItem key={item.title} title={item.title} content={item.content} />
        ))}
      </div>
      <span />
      <div className="address_overview_content__right_items">
        {rightItems.map(item => (
          <AddressOverviewItem key={item.title} title={item.title} content={item.content} />
        ))}
      </div>
    </AddressOverviewContentPanel>
  )
}

const AddressLockScriptItem = ({ title, children }: { title: string; children?: ReactNode }) => {
  return (
    <AddressLockScriptItemPanel>
      <div className="address_lock_script__title">
        <img src={ItemPointIcon} alt="point" />
        <span>{title}</span>
      </div>
      <div className="address_lock_script__content">{children}</div>
    </AddressLockScriptItemPanel>
  )
}

export default (props: React.PropsWithoutRef<RouteComponentProps<{ address: string; hash: string }>>) => {
  const { match, location, history } = props
  const { params } = match
  const { address, hash: lockHash } = params
  const identityHash = address || lockHash
  const { search } = location
  const parsed = queryString.parse(search)
  const { replace } = history

  const page = parsePageNumber(parsed.page, PageParams.PageNo)
  const size = parsePageNumber(parsed.size, PageParams.PageSize)

  const [state, dispatch] = useReducer(reducer, initialState)
  const appContext = useContext(AppContext)

  useEffect(() => {
    if (size > PageParams.MaxPageSize) {
      replace(`/${address ? 'address' : 'lockhash'}/${identityHash}?page=${page}&size=${PageParams.MaxPageSize}`)
    }
    getAddressInfo(identityHash, dispatch)
    getTransactions(identityHash, page, size, dispatch)
    getTipBlockNumber(dispatch)
  }, [replace, identityHash, page, size, dispatch, address])

  const onChange = (pageNo: number, pageSize: number) => {
    replace(`/${address ? 'address' : 'lockhash'}/${identityHash}?page=${pageNo}&size=${pageSize}`)
  }

  const items = [
    {
      title: i18n.t('address.balance'),
      content: `${localeNumberString(shannonToCkb(state.address.balance))} CKB`,
    },
    {
      title: i18n.t('transaction.transactions'),
      content: localeNumberString(state.address.transactions_count),
    },
    {
      title: 'Pending Reward',
      content: '2 blocks',
    },
  ]
  if (state.address.pending_reward_blocks_count) {
    items.push({
      title: i18n.t('address.pending_reward'),
      content: `${state.address.pending_reward_blocks_count} ${
        state.address.pending_reward_blocks_count > 1 ? 'blocks' : 'block'
      }`,
    })
  }
  if (lockHash && state.address) {
    items.push({
      title: `${i18n.t('address.address')} :`,
      content: addressContent(state.address),
    })
  }

  return (
    <AddressContentPanel className="container">
      <AddressHashPanel>
        <div className="address__title">{address ? i18n.t('address.address') : i18n.t('address.lock_hash')}</div>
        <div className="address__hash">{address || lockHash}</div>
        <div
          className="address__copy_iocn"
          role="button"
          tabIndex={-1}
          onKeyDown={() => {}}
          onClick={() => {
            copyElementValue(document.getElementById('address__hash'))
            appContext.toastMessage(i18n.t('common.copied'), 3000)
          }}
        >
          <img src={CopyIcon} alt="copy" />
        </div>
      </AddressHashPanel>
      <AddressTitlePanel>{i18n.t('common.overview')}</AddressTitlePanel>
      <AddressOverviewPanel>
        <AddressOverviewContent items={items} />
        <div className="address_overview__lock_script_title">{`${i18n.t('address.lock_script')} : `}</div>
        <AddressLockScriptItem title={`${i18n.t('address.code_hash')} :`}>
          <span>{state.address.lock_script.code_hash}</span>
        </AddressLockScriptItem>
        <AddressLockScriptItem title={`${i18n.t('address.args')} :`}>
          {state.address.lock_script.args.length === 1 ? (
            <span>{state.address.lock_script.args[0]}</span>
          ) : (
            state.address.lock_script.args.map((arg: string, index: number) => <span>{`#${index}: ${arg}`}</span>)
          )}
        </AddressLockScriptItem>
      </AddressOverviewPanel>
      <AddressTitlePanel>Transactions</AddressTitlePanel>
      <AddressTransactionsPanel>
        {state.transactions &&
          state.transactions.map((transaction: any, index: number) => {
            return (
              transaction &&
              (isMobile() ? (
                <TransactionCard
                  address={state.address.address_hash}
                  confirmation={state.tipBlockNumber - transaction.attributes.block_number + 1}
                  transaction={transaction.attributes}
                  key={transaction.attributes.transaction_hash}
                />
              ) : (
                <TransactionItem
                  address={state.address.address_hash}
                  transaction={transaction.attributes}
                  confirmation={state.tipBlockNumber - transaction.attributes.block_number + 1}
                  key={transaction.attributes.transaction_hash}
                  isEndItem={index === state.transactions.length - 1}
                />
              ))
            )
          })}
        <AddressTransactionsPagition>
          <Pagination
            showQuickJumper
            showSizeChanger
            defaultPageSize={size}
            pageSize={size}
            defaultCurrent={page}
            current={page}
            total={state.total}
            onChange={onChange}
            locale={localeInfo}
          />
        </AddressTransactionsPagition>
      </AddressTransactionsPanel>
    </AddressContentPanel>
  )
}
