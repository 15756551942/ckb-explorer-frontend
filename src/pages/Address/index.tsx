import React, { useEffect, useContext } from 'react'
import { RouteComponentProps } from 'react-router-dom'
import Pagination from 'rc-pagination'
import 'rc-pagination/assets/index.css'
import localeInfo from 'rc-pagination/lib/locale/en_US'
import queryString from 'query-string'
import { AppContext } from '../../contexts/providers/index'
import Content from '../../components/Content'
import TransactionItem from '../../components/Transaction/TransactionItem/index'
import SimpleLabel, { Tooltip } from '../../components/Label'
import {
  AddressContentPanel,
  AddressTitlePanel,
  AddressOverviewPanel,
  AddressScriptContentPanel,
  AddressCommonContent,
  AddressScriptContent,
  AddressScriptLabelPanel,
  AddressTransactionsPanel,
  AddressCommonRowPanel,
  AddressTransactionsPagition,
  ScriptLabelItemPanel,
  ScriptOtherArgs,
  AddressEmptyTransactions,
} from './styled'
import CopyIcon from '../../assets/copy.png'
import BalanceIcon from '../../assets/address_balance.png'
import AddressScriptIcon from '../../assets/address_script.png'
import TransactionsIcon from '../../assets/transactions_green.png'
import ItemPointIcon from '../../assets/item_point.png'
import AddressHashIcon from '../../assets/lock_hash_address.png'
import BlockPendingRewardIcon from '../../assets/block_pending_reward.png'

import { copyElementValue, shannonToCkb } from '../../utils/util'
import { parsePageNumber, startEndEllipsis } from '../../utils/string'
import TransactionCard from '../../components/Transaction/TransactionCard/index'
import { localeNumberString } from '../../utils/number'
import i18n from '../../utils/i18n'
import { isMobile } from '../../utils/screen'
import { StateWithDispatch, AppDispatch, AppActions } from '../../contexts/providers/reducer'
import { PageParams } from '../../utils/const'
import { getAddress } from '../../service/app/address'

const AddressTitle = ({
  address,
  lockHash,
  dispatch,
}: {
  address: string
  lockHash: string
  dispatch: AppDispatch
}) => {
  const identityHash = address || lockHash
  return (
    <AddressTitlePanel>
      <div className="address__title">{address ? i18n.t('address.address') : i18n.t('address.lock_hash')}</div>
      <div className="address__content">
        <code id="address__hash">{identityHash}</code>
        <div
          role="button"
          tabIndex={-1}
          onKeyDown={() => {}}
          onClick={() => {
            copyElementValue(document.getElementById('address__hash'))
            const toast = {
              text: i18n.t('common.copied'),
              timeout: 3000,
            }
            dispatch({
              type: AppActions.ShowToastMessage,
              payload: {
                ...toast,
              },
            })
          }}
        >
          <img src={CopyIcon} alt="copy" />
        </div>
      </div>
    </AddressTitlePanel>
  )
}

const AddressOverview = ({ value }: { value: string }) => {
  return <AddressOverviewPanel>{value}</AddressOverviewPanel>
}

const ScriptLabelItem = ({ name, value, noIcon = false }: { name: string; value: string; noIcon?: boolean }) => {
  return (
    <ScriptLabelItemPanel>
      {!noIcon && <img src={ItemPointIcon} alt="item point" />}
      <div>{name}</div>
      <code>{value}</code>
    </ScriptLabelItemPanel>
  )
}

const AddressScriptLabel = ({ image, label, script }: { image: string; label: string; script: State.Script }) => {
  return (
    <div>
      <AddressScriptLabelPanel>
        <img src={image} alt="script" />
        <span>{label}</span>
      </AddressScriptLabelPanel>
      <AddressScriptContentPanel>
        <AddressScriptContent>
          {script.code_hash && <ScriptLabelItem name={`${i18n.t('address.code_hash')} :`} value={script.code_hash} />}
          {script.args.length === 1 ? (
            <ScriptLabelItem name={`${i18n.t('address.args')} :`} value={script.args[0]} />
          ) : (
            script.args.map((arg: string, index: number) => {
              return index === 0 ? (
                <ScriptLabelItem name={`${i18n.t('address.args')} :`} value={`#${index}: ${arg}`} />
              ) : (
                <ScriptOtherArgs>
                  <ScriptLabelItem name="" value={`#${index}: ${arg}`} noIcon />
                </ScriptOtherArgs>
              )
            })
          )}
        </AddressScriptContent>
      </AddressScriptContentPanel>
    </div>
  )
}

const PendingRewardTooltip: Tooltip = {
  tip: i18n.t('address.pending_reward_tooltip'),
  haveHelpIcon: true,
  offset: 0.7,
}

const addressContent = (address: State.Address) => {
  const addressText = isMobile() ? startEndEllipsis(address.address_hash, 10) : address.address_hash
  return address.address_hash ? addressText : i18n.t('address.unable_decode_address')
}

export const Address = ({
  dispatch,
  history: { replace },
  location: { search },
  match: { params },
}: React.PropsWithoutRef<StateWithDispatch & RouteComponentProps<{ address: string; hash: string }>>) => {
  const { address, hash: lockHash } = params
  const identityHash = address || lockHash
  const parsed = queryString.parse(search)

  const page = parsePageNumber(parsed.page, PageParams.PageNo)
  const size = parsePageNumber(parsed.size, PageParams.PageSize)

  const { addressState, tipBlockNumber } = useContext(AppContext)

  useEffect(() => {
    if (size > PageParams.MaxPageSize) {
      replace(`/${address ? 'address' : 'lockhash'}/${identityHash}?page=${page}&size=${PageParams.MaxPageSize}`)
    }
    getAddress(identityHash, page, size, dispatch)
  }, [replace, identityHash, page, size, dispatch, address])

  const onChange = (pageNo: number, pageSize: number) => {
    replace(`/${address ? 'address' : 'lockhash'}/${identityHash}?page=${pageNo}&size=${pageSize}`)
  }

  return (
    <Content>
      <AddressContentPanel className="container">
        <AddressTitle address={address} lockHash={lockHash} dispatch={dispatch} />
        <AddressOverview value={i18n.t('common.overview')} />
        <AddressCommonContent>
          <AddressCommonRowPanel>
            <SimpleLabel
              image={BalanceIcon}
              label={`${i18n.t('address.balance')} : `}
              value={`${localeNumberString(shannonToCkb(addressState.address.balance))} CKB`}
            />
            <SimpleLabel
              image={TransactionsIcon}
              label={`${i18n.t('transaction.transactions')} : `}
              value={localeNumberString(addressState.address.transactions_count)}
            />
          </AddressCommonRowPanel>
          {addressState.address.pending_reward_blocks_count ? (
            <SimpleLabel
              image={BlockPendingRewardIcon}
              label={`${i18n.t('address.pending_reward')} : `}
              value={`${addressState.address.pending_reward_blocks_count} 
                ${addressState.address.pending_reward_blocks_count > 1 ? 'blocks' : 'block'}`}
              tooltip={PendingRewardTooltip}
            />
          ) : null}
          {lockHash && addressState.address && (
            <SimpleLabel
              image={AddressHashIcon}
              label={`${i18n.t('address.address')} :`}
              value={addressContent(addressState.address)}
            />
          )}
          <AddressScriptLabel
            image={AddressScriptIcon}
            label={`${i18n.t('address.lock_script')} : `}
            script={addressState.address.lock_script}
          />
        </AddressCommonContent>

        {addressState.transactions && addressState.transactions.length > 0 ? (
          <AddressTransactionsPanel>
            <AddressOverview value={i18n.t('transaction.transactions')} />
            <div>
              {addressState.transactions.map((transaction: any) => {
                return (
                  transaction &&
                  (isMobile() ? (
                    <TransactionCard
                      address={addressState.address.address_hash}
                      confirmation={tipBlockNumber - transaction.attributes.block_number + 1}
                      transaction={transaction.attributes}
                      key={transaction.attributes.transaction_hash}
                    />
                  ) : (
                    <TransactionItem
                      address={addressState.address.address_hash}
                      transaction={transaction.attributes}
                      confirmation={tipBlockNumber - transaction.attributes.block_number + 1}
                      key={transaction.attributes.transaction_hash}
                    />
                  ))
                )
              })}
            </div>
            <AddressTransactionsPagition>
              <Pagination
                showQuickJumper
                showSizeChanger
                defaultPageSize={size}
                pageSize={size}
                defaultCurrent={page}
                current={page}
                total={addressState.total}
                onChange={onChange}
                locale={localeInfo}
              />
            </AddressTransactionsPagition>
          </AddressTransactionsPanel>
        ) : (
          <AddressEmptyTransactions />
        )}
      </AddressContentPanel>
    </Content>
  )
}

export default Address
