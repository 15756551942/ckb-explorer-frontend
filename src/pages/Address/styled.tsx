import styled from 'styled-components'
import { CommonPagition } from '../BlockList/styled'

export const AddressContentPanel = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 40px;
  margin-bottom: 40px;
  width: 100%;

  @media (max-width: 700px) {
    margin-top: 30px;
    padding: 0px 20px 20px 20px;
  }
`

export const AddressPendingRewardTitlePanel = styled.div`
  display: flex;
  align-items: center;

  .address__pending_reward_help {
    margin-left: 20px;
    transform: translateY(2px);
    position: relative;
    > img {
      width: 20px;
      height: 20px;
    }
  }
`

export const AddressLockScriptPanel = styled.div`
  width: 100%;
  .address__lock_script_title {
    font-weight: 500;
  }
`

export const AddressLockScriptItemPanel = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  align-items: flex-start;
  margin-top: 10px;

  @media (max-width: 700px) {
    flex-direction: column;
  }

  .address_lock_script__title {
    > img {
      width: 9px;
      height: 9px;

      @media (max-width: 700px) {
        width: 5px;
        height: 5px;
        transform: translateY(-2px);
      }
    }

    > span {
      margin-left: 10px;
      font-weight: 500;

      @media (max-width: 700px) {
        margin-left: 5px;
      }
    }
  }

  .address_lock_script__content {
    flex: 1;
    margin-left: 20px;
    display: flex;
    flex-direction: column;

    @media (max-width: 700px) {
      margin-left: 10px;
      word-wrap: break-word;
      word-break: break-all;
    }
  }
`

export const AddressTransactionsPanel = styled.div`
  width: 100%;
`

export const AddressTransactionsPagition = styled(CommonPagition)`
  margin: 40px 0 0px 0;
  width: 100%;

  @media (max-width: 700px) {
    margin: 20px 0 30px 3%;
  }
`
