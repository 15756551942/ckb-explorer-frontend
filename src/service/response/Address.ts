import { Script } from './Script'

export interface Address {
  address_hash: string
  lock_hash: string
  balance: number
  transactions_count: number
  cell_consumed: number
  lock_script: Script
  pending_reward_blocks_count: number
}

export interface AddressWrapper {
  id: number
  type: string
  attributes: Address
}
