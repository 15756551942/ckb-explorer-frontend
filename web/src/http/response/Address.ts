import Script from './Script'

export default interface Address {
  address_hash: string
  balance: number
  transactions_count: number
  cell_consumed: number
  lock_script: Script
}
