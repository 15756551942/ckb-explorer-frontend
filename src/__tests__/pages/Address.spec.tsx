import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Address from '../../pages/Address'

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    length: 13,
    push: jest.fn(),
    block: jest.fn(),
    createHref: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    liten: jest.fn(),
    replace: jest.fn(),
    action: 'REPLACE',
    location: null,
  }),
  useLocation: () => ({
    pathname: '/welcome',
    hash: '',
    search: '',
    state: '',
  }),
  useParams: jest.fn().mockReturnValue({ id: '123' }),
}))

describe('Active Page', () => {
  it('is defined', () => {
    const app = shallow(<Address />)
    expect(app).toBeDefined()
  })

  it('shallow renders with empty data', async () => {
    const wrapper = renderer.create(<Address />).toJSON()
    expect(wrapper).toMatchSnapshot()
  })

  it('shallow renders', async () => {
    const wrapper = shallow(<Address />)
    expect(wrapper).toMatchSnapshot()
  })
})
