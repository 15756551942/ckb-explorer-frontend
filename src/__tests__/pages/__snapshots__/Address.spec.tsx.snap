// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active Page shallow renders 1`] = `
<_default>
  <styled.div
    className="container"
  >
    <_default
      specialAddress=""
      title="Address"
    >
      <AddressTitleOverview />
    </_default>
    <AddressAssetCompState />
    <AddressStateTransactions
      currentPage={1}
      pageSize={10}
    />
  </styled.div>
</_default>
`;

exports[`Active Page shallow renders with empty data 1`] = `
<div
  className="sc-fznxsB cFeRni"
>
  <div
    className="sc-fznWqX bNxGGa container"
  >
    <div
      className="sc-AxiKw cszbrx"
    >
      <div
        className="hash__card__content__panel"
        id="hash_content"
      >
        <div
          className="hash__title"
        >
          Address
        </div>
        <div
          className="hash__card__hash__content"
        >
          <div
            id="hash__text"
          >
            <span
              className="monospace"
            >
              
            </span>
          </div>
          <div
            className="sc-AxhUy ifPyXr hash__copy_icon"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseOver={[Function]}
            role="button"
            tabIndex={-1}
          >
            <img
              alt="copy"
              src="copy.png"
            />
          </div>
        </div>
        
        <div
          className="monospace"
          id="hash__value"
        />
      </div>
      <div
        className="sc-fznJRM eAGdwR"
      >
        <div
          className="sc-fzqBkg cTxTQP"
        >
          <div
            className="overview__separate"
          />
          <div
            className="sc-fzqPZZ gwMzMo"
          >
            <div
              className="overview_content__left_items"
            >
              <div
                className="sc-fzoxKX cGEWAH"
              >
                <div
                  className="overview_item__title__panel"
                >
                  <div
                    className="overview_item__title"
                  >
                    Live Cells
                  </div>
                </div>
                <div
                  className="overview_item__value"
                >
                  0
                </div>
              </div>
            </div>
            <div
              className="overview_content__right_items"
            >
              <div
                className="sc-fzoxKX cGEWAH"
              >
                <div
                  className="overview_item__title__panel"
                >
                  <div
                    className="overview_item__title"
                  >
                    Block Mined
                  </div>
                </div>
                <div
                  className="overview_item__value"
                >
                  0
                </div>
              </div>
            </div>
          </div>
          <div
            className="sc-fzqARJ dXYecO"
            onClick={[Function]}
            onKeyUp={[Function]}
            role="button"
            tabIndex={0}
          >
            <div>
              Lock Script
            </div>
            <img
              alt="lock script"
              src="arrow_down.png"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
